### PR TITLE
sidecar: Make it possible to pass command arguments to scyla

### LINKED ISSUE #177: Add ability to modify/replace default SCYLLA_ARGS
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**

Scylla Operator has support for modifying `scylla.yaml` but not for modifying command line arguments for `scylla` binary.

These arguments are configured by `/etc/sysconfig/scylla-server` file:

```
...
# scylla arguments
SCYLLA_ARGS="--log-to-syslog 0 --log-to-stdout 1 --default-log-level info --network-stack posix"
...
```

It'll be useful to have ability to modify them too.

**What is use case behind this feature:**

SCT has a feature in test configuration to modify these arguments and some of the tests use it. Even more, SCT appends following options by default: `--blocked-reactor-notify-ms 500 --abort-on-lsa-bad-alloc 1 --abort-on-seastar-bad-alloc --abort-on-
internal-error 1 --abort-on-ebadf 1 --enable-sstable-key-validation 1`. So, for making k8s backend complete we need to implement this feature.

### PR DESCRIPTION
**Pull request by dkropachev**
Addresses https://github.com/scylladb/scylla-operator/issues/177

**Description of your changes:**
Exposes scyllaArgs at cluster.spec 

Using scyllaArgs will make scylla startup to fail on versions that does not contain https://github.com/scylladb/scylla/pull/7458

**Which issue is resolved by this Pull Request:**
Resolves #177

**Checklist:**
- [ ] ~~Documentation has been updated, if necessary.~~
- [x] Image has been built (`make docker-build`) on the last commit.

### CODE REVIEW

**Comment by zimnx** [comment]
> Using scyllaArgs will make scylla startup to fail on versions that does not contain [scylladb/scylla#7458](https://github.com/scylladb/scylla/pull/7458)

Which version of Scylla is going to support it? 
Either we should add validation if used Scylla versions supports that, or at least log warning that it may not work.

**Comment by dkropachev** [comment]
> > Using scyllaArgs will make scylla startup to fail on versions that does not contain [scylladb/scylla#7458](https://github.com/scylladb/scylla/pull/7458)
> 
> Which version of Scylla is going to support it?
> Either we should add validation if used Scylla versions supports that, or at least log warning that it may not work.

Hopefully it will work on 4.3

File: pkg/controllers/sidecar/config/config.go [hunk 1]
```diff
@@ -178,17 +179,22 @@ func (s *ScyllaConfig) setupEntrypoint(ctx context.Context) (*exec.Cmd, error) {
...
+	seedsStr := strings.Join(seeds, ",")
+	overprovisionedStr := "0"
+	smpStr := strconv.Itoa(shards)
+	args := map[string]*string{
+		"listen-address":        &m.IP,
+		"broadcast-address":     &m.StaticIP,
+		"broadcast-rpc-address": &m.StaticIP,
+		"seeds":                 &seedsStr,
+		"developer-mode":        &devMode,
+		"overprovisioned":       &overprovisionedStr,
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply pkg/controllers/sidecar/config/config.go:192]: I have switched to "k8s.io/utils/pointer", since we are already have it.

File: pkg/controllers/sidecar/config/config.go [hunk 2]
```diff
@@ -203,10 +209,39 @@ func (s *ScyllaConfig) setupEntrypoint(ctx context.Context) (*exec.Cmd, error) {
 		if err := s.validateCpuSet(ctx, cpusAllowed, shards); err != nil {
 			return nil, errors.WithStack(err)
 		}
-		args = append(args, fmt.Sprintf("--cpuset=%s", cpusAllowed))
+		args["cpuset"] = &cpusAllowed
+	}
+
+	if cluster.Spec.ScyllaArgs != nil {
+		tmp := regexp.MustCompile(`--([^= ]+)(="[^"]+"|=[^ ]+|[ \t]+"[^"]+"|[ \t]+[^-][^-]?[^ ]*|)`).FindAllStringSubmatch(*cluster.Spec.ScyllaArgs, -1)
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply pkg/controllers/sidecar/config/config.go:216]: switched ScyllaArgs to map

[the remark this answers is not in the document]
Comment (dkropachev) [reply pkg/controllers/sidecar/config/config.go:216]: > If you make ScyllaArgs a slice, you won't have to worry about slicing.
> For validation we have a validating webhook which rejects CRD with invalid fields. Check `cluster_webhook.go`
Problem here that you don't have list of valid fields, unless you run scylla --help.

File: pkg/api/v1alpha1/cluster_types.go [hunk 3]
```diff
@@ -51,7 +51,8 @@ type ClusterSpec struct {
 	// Sysctl properties to be applied during initialization
 	// given as a list of key=value pairs.
 	// Example: fs.aio-max-nr=232323
-	Sysctls []string `json:"sysctls,omitempty"`
+	Sysctls    []string `json:"sysctls,omitempty"`
+	ScyllaArgs *string  `json:"scyllaArgs,omitempty"`
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply pkg/api/v1alpha1/cluster_types.go]: It would be better to have it as Map, but unfortunatelly controller-get v0.3.0 does not support nillable values in the map, whcih does not allow us to have flag arguments there.
Currently I've move it back to string.

File: pkg/controllers/sidecar/config/config.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (dkropachev) [reply pkg/controllers/sidecar/config/config.go]: Turned out that controller-get v0.3.0 does not support nillable values in the map, which does not allow us to have flag arguments there, so I have move it back to string and add tests to that part of the logic.

File: pkg/controllers/sidecar/config/config.go [hunk 4]
```diff
@@ -4,8 +4,10 @@ import (
 	"context"
 	"fmt"
 	"io/ioutil"
+	pointer2 "k8s.io/utils/pointer"
```
Comment (zimnx) [inline pkg/controllers/sidecar/config/config.go:7]: pointer2? package is called `pointer`, i don't see name conflict in mports

File: pkg/api/v1alpha1/cluster_types.go [hunk 5]
```diff
@@ -51,7 +51,8 @@ type ClusterSpec struct {
 	// Sysctl properties to be applied during initialization
 	// given as a list of key=value pairs.
 	// Example: fs.aio-max-nr=232323
-	Sysctls []string `json:"sysctls,omitempty"`
+	Sysctls    []string `json:"sysctls,omitempty"`
+	ScyllaArgs string   `json:"scyllaArgs,omitempty"`
```
Comment (zimnx) [inline pkg/api/v1alpha1/cluster_types.go:55]: now it's required, but should be optional

File: pkg/controllers/sidecar/config/config_test.go [hunk 6]
```diff
@@ -186,3 +186,59 @@ func writeTempFile(t *testing.T, namePattern, content string) string {
 
 	return tmp.Name()
 }
+
+func TestScyllaArguments(t *testing.T) {
```
Comment (zimnx) [inline pkg/controllers/sidecar/config/config_test.go:190]: Split it to multiple tests. Failure in one case shouldn't cause failure in cases where regexp works fine.
You can use table tests: https://github.com/golang/go/wiki/TableDrivenTests

File: examples/eks/manager.yaml [hunk 7]
```diff
@@ -0,0 +1,275 @@
+apiVersion: v1
```
Comment (zimnx) [inline examples/eks/manager.yaml:1]: master is missing this update after manager merge. Please extract it to separate commit

File: pkg/controllers/sidecar/config/config.go [hunk 8]
```diff
@@ -156,6 +158,22 @@ func loadProperties(fileName string, logger log.Logger) *properties.Properties {
 	return p
 }
 
+var scyllaArgumentsRegexp = regexp.MustCompile(`--([^= ]+)(="[^"]+"|=[^ ]+|[ \t]+"[^"]+"|[ \t]+[^-][^-]?[^ ]*|)`)
+
+func convertScyllaArguments(scyllaArguments string) map[string]*string {
+	output := make(map[string]*string)
```
Comment (zimnx) [inline pkg/controllers/sidecar/config/config.go:164]: nil pointers aren't added to this map, you can keep `map[string]string`. Later there's unnecessary dereferece:  `pointer2.StringPtr(strings.TrimSpace(*argValue))`

File: examples/eks/manager.yaml (hunk 7 above)
Comment (dkropachev) [reply examples/eks/manager.yaml:1]: removed

File: pkg/controllers/sidecar/config/config.go (hunk 8 above)
Comment (dkropachev) [reply pkg/controllers/sidecar/config/config.go:164]: fixed

File: pkg/controllers/sidecar/config/config_test.go (hunk 6 above)
Comment (dkropachev) [reply pkg/controllers/sidecar/config/config_test.go:190]: done

File: pkg/api/v1alpha1/cluster_types.go (hunk 5 above)
Comment (dkropachev) [reply pkg/api/v1alpha1/cluster_types.go:55]: scylla.scylladb.com_clusters.yaml resulted in 
```
          required:
          - agentVersion
          - datacenter
          - version

```
So it is not required on scheme level, also I've checked it by deploying cluster with no scyllaArgs and it passed thru

File: pkg/controllers/sidecar/config/config.go (hunk 4 above)
Comment (dkropachev) [reply pkg/controllers/sidecar/config/config.go]: fixed

File: pkg/controllers/sidecar/config/config.go [hunk 9]
```diff
@@ -203,10 +222,29 @@ func (s *ScyllaConfig) setupEntrypoint(ctx context.Context) (*exec.Cmd, error) {
 		if err := s.validateCpuSet(ctx, cpusAllowed, shards); err != nil {
 			return nil, errors.WithStack(err)
 		}
-		args = append(args, fmt.Sprintf("--cpuset=%s", cpusAllowed))
+		args["cpuset"] = &cpusAllowed
+	}
+
+	if len(cluster.Spec.ScyllaArgs) > 0 {
+		for argName, argValue := range convertScyllaArguments(cluster.Spec.ScyllaArgs) {
```
Comment (zimnx) [inline pkg/controllers/sidecar/config/config.go:229]: It doesn't work for older Scylla versions, it doesn't even work for released versions. We definitely need some kind of warning or info log level message about missing support.

Comment (dkropachev) [reply pkg/controllers/sidecar/config/config.go]: Validation has been added.

File: pkg/api/v1alpha1/cluster_types.go [hunk 10]
```diff
@@ -268,6 +270,12 @@ type ClusterList struct {
 	Items           []Cluster `json:"items"`
 }
 
+// Version of scylla docker starting from which passing arguments via entry-point is supported
+var ScyllaVersionThatSupportsArgsText = "4.2.0"
+var ScyllaVersionThatSupportsArgs *semver.Version
```
Comment (zimnx) [inline pkg/api/v1alpha1/cluster_types.go:275]: why pointer? Every usage dereference it. Use value.

**Review by zimnx — COMMENTED** [review]
last comments

File: pkg/api/v1alpha1/cluster_types.go [hunk 11]
```diff
@@ -268,6 +270,12 @@ type ClusterList struct {
...
+
 func init() {
+	tmp, _ := semver.Parse(ScyllaVersionThatSupportsArgsText)
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply pkg/api/v1alpha1/cluster_types.go]: fixed

File: pkg/api/v1alpha1/cluster_types.go (hunk 10 above)
Comment (dkropachev) [reply pkg/api/v1alpha1/cluster_types.go]: fixed

**Review by zimnx — APPROVED** [review]
LGTM
